File: backend/app/schemas/evaluations.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class CreateEvaluationSetRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank")
        return trimmed

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("description must not be blank")
        return trimmed


class EvaluationSetResponse(BaseModel):
    evaluation_set_id: str
    name: str
    description: str | None = None
    question_count: int = 0
    created_at: datetime
    updated_at: datetime


class EvaluationSetListResponse(BaseModel):
    items: list[EvaluationSetResponse]
    total: int
    limit: int
    offset: int


class CreateEvaluationQuestionRequest(BaseModel):
    question: str = Field(min_length=1, max_length=8000)
    expected_answer: str | None = Field(default=None, max_length=8000)
    expected_document_id: str | None = None
    expected_page_number: int | None = Field(default=None, ge=1)
    tags: list[str] = Field(default_factory=list, max_length=50)
    metadata: dict[str, object] = Field(default_factory=dict)

    @field_validator("question")
    @classmethod
    def validate_question(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("question must not be blank")
        return trimmed

    @field_validator("expected_answer")
    @classmethod
    def validate_expected_answer(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("expected_answer must not be blank")
        return trimmed

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, value: list[str]) -> list[str]:
        normalized_tags: list[str] = []
        for tag in value:
            trimmed = tag.strip()
            if not trimmed:
                raise ValueError("tags must not contain blank values")
            normalized_tags.append(trimmed)
        return normalized_tags


class EvaluationQuestionResponse(BaseModel):
    evaluation_question_id: str
    evaluation_set_id: str
    question: str
    expected_answer: str | None = None
    expected_document_id: str | None = None
    expected_page_number: int | None = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, object] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class EvaluationQuestionListResponse(BaseModel):
    evaluation_set_id: str
    items: list[EvaluationQuestionResponse]
    total: int
    limit: int
    offset: int


class EvaluationRunConfig(BaseModel):
    top_k: int | None = Field(default=None, ge=1, le=200)
    rerank: bool = True
    model_name: str | None = Field(default=None, min_length=3, max_length=128)
    selected_document_ids: list[str] = Field(default_factory=list, max_length=50)
    metric_options: dict[str, bool | int | float | str] = Field(default_factory=dict)

    @field_validator("model_name")
    @classmethod
    def validate_model_name(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("model_name must not be blank")
        return trimmed

    @field_validator("selected_document_ids")
    @classmethod
    def validate_selected_document_ids(cls, value: list[str]) -> list[str]:
        normalized_ids: list[str] = []
        for document_id in value:
            trimmed = document_id.strip()
            if not trimmed:
                raise ValueError("selected_document_ids must not contain blank values")
            normalized_ids.append(trimmed)
        return normalized_ids


class RunEvaluationRequest(BaseModel):
    evaluation_set_id: str = Field(min_length=3, max_length=64)
    config: EvaluationRunConfig = Field(default_factory=EvaluationRunConfig)

    @field_validator("evaluation_set_id")
    @classmethod
    def validate_evaluation_set_id(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("evaluation_set_id must not be blank")
        return trimmed


class RunEvaluationResponse(BaseModel):
    evaluation_run_id: str
    status: Literal["queued"] = "queued"


class EvaluationStatusResponse(BaseModel):
    evaluation_run_id: str
    status: Literal["queued", "running", "completed", "failed"]
    score: float | None = None
    updated_at: datetime | None = None


class EvaluationRunResultResponse(BaseModel):
    evaluation_result_id: str
    evaluation_question_id: str
    question: str
    status: str
    generated_answer: str | None = None
    retrieval_score: float | None = None
    faithfulness_score: float | None = None
    citation_accuracy_score: float | None = None
    answer_relevance_score: float | None = None
    latency_ms: int | None = None
    metrics: dict[str, object] = Field(default_factory=dict)
    failure_reason: str | None = None
    failure_type: str | None = None
    details: dict[str, object] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class EvaluationRunResultListResponse(BaseModel):
    items: list[EvaluationRunResultResponse]
    total: int
    limit: int
    offset: int


class EvaluationRunDetailResponse(BaseModel):
    evaluation_run_id: str
    evaluation_set_id: str
    status: Literal["queued", "running", "completed", "failed"]
    config: dict[str, object] = Field(default_factory=dict)
    summary: dict[str, object] | None = None
    failure_reason: str | None = None
    failure_type: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    results: EvaluationRunResultListResponse

File: frontend/src/lib/api/evaluations.ts
import { apiRequest } from "@/lib/api/request";
import type { components } from "@/lib/api/generated/schema";
import type { ChunkingProfileConfigInput } from "@/lib/schemas/chunking-profiles";

type Schemas = components["schemas"];

export type EvaluationSetResponse = Omit<
  Schemas["EvaluationSetResponse"],
  "status" | "version" | "owner_id" | "scope"
> & {
  status: string;
  version: number;
  owner_id?: string | null;
  scope: Record<string, unknown>;
};
export type EvaluationSetListResponse = {
  items: EvaluationSetResponse[];
  total: number;
  limit: number;
  offset: number;
};
export type CreateEvaluationSetRequest = Schemas["CreateEvaluationSetRequest"];
export type EvaluationQuestionResponse = Omit<
  Schemas["EvaluationQuestionResponse"],
  "tags" | "difficulty" | "owner_id"
> & {
  tags: string[];
  difficulty?: string | null;
  owner_id?: string | null;
};
export type EvaluationQuestionListResponse = {
  evaluation_set_id: string;
  items: EvaluationQuestionResponse[];
  total: number;
  limit: number;
  offset: number;
};
export type CreateEvaluationQuestionRequest =
  Schemas["CreateEvaluationQuestionRequest"];
type BaseEvaluationRunConfig = Schemas["EvaluationRunConfig"];
type BaseRunEvaluationRequest = Schemas["RunEvaluationRequest"];
export type RunEvaluationResponse = Schemas["RunEvaluationResponse"];
export type EvaluationRunResultResponse =
  Schemas["EvaluationRunResultResponse"];
export type EvaluationRunDetailResponse =
  Schemas["EvaluationRunDetailResponse"];

export type EvaluationChunkingComparisonTarget = {
  label?: string | null;
  chunking_profile_id?: string | null;
  chunking_profile_config?: ChunkingProfileConfigInput | null;
};

export type EvaluationRegressionThresholds = {
  retrieval_hit_rate_min?: number | null;
  citation_accuracy_score_min?: number | null;
  faithfulness_score_min?: number | null;
  max_not_found_rate?: number | null;
};

export type EvaluationRunConfig = Omit<
  BaseEvaluationRunConfig,
  | "chunking_profile_id"
  | "chunking_profile_config"
  | "comparison_targets"
  | "regression_thresholds"
  | "run_name"
> & {
  run_name?: string | null;
  chunking_profile_id?: string | null;
  chunking_profile_config?: ChunkingProfileConfigInput | null;
  comparison_targets?: EvaluationChunkingComparisonTarget[];
  regression_thresholds?: EvaluationRegressionThresholds | null;
  /** UUID of an OrgModelProfile to use for this run. When omitted the org's evaluations task profile is resolved at runtime. */
  model_profile_id?: string | null;
};

export type RunEvaluationRequest = Omit<BaseRunEvaluationRequest, "config"> & {
  config?: EvaluationRunConfig;
};

export type UpdateEvaluationSetRequest = {
  name?: string | null;
  description?: string | null;
  scope?: Record<string, unknown> | null;
};

export type UpdateEvaluationQuestionRequest = {
  question?: string | null;
  expected_answer?: string | null;
  expected_document_id?: string | null;
  expected_page_number?: number | null;
  difficulty?: "easy" | "medium" | "hard" | null;
  tags?: string[] | null;
  metadata?: Record<string, unknown> | null;
};

export type ImportCasesRequest = {
  format: "json" | "csv";
  data: string;
  skip_duplicates?: boolean;
};

export type ImportCasesResponse = {
  imported: number;
  skipped_duplicates: number;
  validation_errors: string[];
};

export type PublishDatasetResponse = {
  evaluation_set_id: string;
  version_number: number;
  question_count: number;
  status: "published";
};

export type DuplicateDatasetResponse = {
  evaluation_set_id: string;
  name: string;
  question_count: number;
  status: "draft";
  created_at: string;
};

export type DatasetValidationIssue = {
  evaluation_question_id: string;
  question_preview: string;
  issue_type:
    | "missing_scope"
    | "deleted_source"
    | "inaccessible_document"
    | "no_expected_answer"
    | "duplicate";
  detail: string;
};

export type ValidateDatasetResponse = {
  evaluation_set_id: string;
  is_valid: boolean;
  issue_count: number;
  issues: DatasetValidationIssue[];
};

export type EvaluationDatasetVersionResponse = {
  version_id: string;
  evaluation_set_id: string;
  version_number: number;
  question_count: number;
  published_by_id: string | null;
  published_at: string | null;
  created_at: string;
};

export type EvaluationDatasetVersionListResponse = {
  evaluation_set_id: string;
  items: EvaluationDatasetVersionResponse[];
  total: number;
};

export type ConvertFeedbackToCasesRequest = {
  evaluation_set_id: string;
  feedback_ids: string[];
  default_difficulty?: "easy" | "medium" | "hard" | null;
};

export type ConvertFeedbackToCasesResponse = {
  created: number;
  skipped: number;
  evaluation_set_id: string;
};

export async function createEvaluationSet(
  payload: CreateEvaluationSetRequest,
): Promise<EvaluationSetResponse> {
  return apiRequest<EvaluationSetResponse>("/evaluation-sets", {
    method: "POST",
    json: payload,
  });
}

export async function listEvaluationSets(
  params: { limit?: number; offset?: number } = {},
): Promise<EvaluationSetListResponse> {
  return apiRequest<EvaluationSetListResponse>("/evaluation-sets", {
    query: {
      limit: params.limit,
      offset: params.offset,
    },
  });
}

export async function updateEvaluationSet(
  evaluationSetId: string,
  payload: UpdateEvaluationSetRequest,
): Promise<EvaluationSetResponse> {
  return apiRequest<EvaluationSetResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}`,
    {
      method: "PATCH",
      json: payload,
    },
  );
}

export async function deleteEvaluationSet(
  evaluationSetId: string,
): Promise<void> {
  return apiRequest<void>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}`,
    { method: "DELETE" },
  );
}

export async function publishEvaluationSet(
  evaluationSetId: string,
): Promise<PublishDatasetResponse> {
  return apiRequest<PublishDatasetResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/publish`,
    { method: "POST" },
  );
}

export async function duplicateEvaluationSet(
  evaluationSetId: string,
): Promise<DuplicateDatasetResponse> {
  return apiRequest<DuplicateDatasetResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/duplicate`,
    { method: "POST" },
  );
}

export async function importEvaluationCases(
  evaluationSetId: string,
  payload: ImportCasesRequest,
): Promise<ImportCasesResponse> {
  return apiRequest<ImportCasesResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/import`,
    {
      method: "POST",
      json: payload,
    },
  );
}

export async function validateEvaluationDataset(
  evaluationSetId: string,
): Promise<ValidateDatasetResponse> {
  return apiRequest<ValidateDatasetResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/validate`,
  );
}

export async function listDatasetVersions(
  evaluationSetId: string,
): Promise<EvaluationDatasetVersionListResponse> {
  return apiRequest<EvaluationDatasetVersionListResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/versions`,
  );
}

export async function convertFeedbackToCases(
  payload: ConvertFeedbackToCasesRequest,
): Promise<ConvertFeedbackToCasesResponse> {
  return apiRequest<ConvertFeedbackToCasesResponse>(
    "/evaluation-sets/from-feedback",
    {
      method: "POST",
      json: payload,
    },
  );
}

export async function createEvaluationQuestion(
  evaluationSetId: string,
  payload: CreateEvaluationQuestionRequest,
): Promise<EvaluationQuestionResponse> {
  return apiRequest<EvaluationQuestionResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/questions`,
    {
      method: "POST",
      json: payload,
    },
  );
}

export async function listEvaluationQuestions(
  evaluationSetId: string,
  params: { limit?: number; offset?: number } = {},
): Promise<EvaluationQuestionListResponse> {
  return apiRequest<EvaluationQuestionListResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/questions`,
    {
      query: {
        limit: params.limit,
        offset: params.offset,
      },
    },
  );
}

export async function updateEvaluationQuestion(
  evaluationSetId: string,
  evaluationQuestionId: string,
  payload: UpdateEvaluationQuestionRequest,
): Promise<EvaluationQuestionResponse> {
  return apiRequest<EvaluationQuestionResponse>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/questions/${encodeURIComponent(evaluationQuestionId)}`,
    {
      method: "PATCH",
      json: payload,
    },
  );
}

export async function deleteEvaluationQuestion(
  evaluationSetId: string,
  evaluationQuestionId: string,
): Promise<void> {
  return apiRequest<void>(
    `/evaluation-sets/${encodeURIComponent(evaluationSetId)}/questions/${encodeURIComponent(evaluationQuestionId)}`,
    { method: "DELETE" },
  );
}

export async function runEvaluation(
  payload: RunEvaluationRequest,
): Promise<RunEvaluationResponse> {
  return apiRequest<RunEvaluationResponse>("/evaluations/run", {
    method: "POST",
    json: payload,
  });
}

export async function getEvaluationRun(
  evaluationRunId: string,
  params: { limit?: number; offset?: number } = {},
): Promise<EvaluationRunDetailResponse> {
  return apiRequest<EvaluationRunDetailResponse>(
    `/evaluations/runs/${encodeURIComponent(evaluationRunId)}`,
    {
      query: {
        limit: params.limit,
        offset: params.offset,
      },
    },
  );
}

export type LocalModelMetrics = {
  invalid_json_rate: number | null;
  timeout_rate: number | null;
  fallback_frequency: number | null;
  estimated_compute_latency_ms: number | null;
  tokens_per_second: number | null;
};

export type EvaluationRunSummaryResponse = {
  evaluation_run_id: string;
  evaluation_set_id: string;
  run_name: string | null;
  status: string;
  summary: Record<string, unknown> | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  model_profile_key: string | null;
  provider_type: string | null;
  provider_profile: string | null;
};

export type EvaluationRunListResponse = {
  items: EvaluationRunSummaryResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type MetricDelta = {
  metric: string;
  label: string;
  run_a_value: number | null;
  run_b_value: number | null;
  delta: number | null;
  is_regression: boolean;
  is_improvement: boolean;
};

export type CaseComparisonRow = {
  evaluation_question_id: string;
  question: string;
  difficulty: string | null;
  tags: string[];
  run_a: EvaluationRunResultResponse | null;
  run_b: EvaluationRunResultResponse | null;
  regression: boolean;
  improvement: boolean;
};

export type RunComparisonResponse = {
  run_a: EvaluationRunSummaryResponse;
  run_b: EvaluationRunSummaryResponse;
  metric_deltas: MetricDelta[];
  regression_count: number;
  improvement_count: number;
  cases: CaseComparisonRow[];
  total_cases: number;
  filters_applied: Record<string, unknown>;
};

export type CompareRunsParams = {
  difficulty?: string | null;
  tags?: string | null;
  case_status?: "all" | "regression" | "improvement" | "failed_any";
  failure_type?: string | null;
};

export async function listEvaluationRuns(
  params: {
    evaluation_set_id?: string;
    status?: string;
    limit?: number;
    offset?: number;
  } = {},
): Promise<EvaluationRunListResponse> {
  return apiRequest<EvaluationRunListResponse>("/evaluations/runs", {
    query: {
      evaluation_set_id: params.evaluation_set_id,
      status: params.status,
      limit: params.limit,
      offset: params.offset,
    },
  });
}

export async function compareEvaluationRuns(
  runAId: string,
  runBId: string,
  params: CompareRunsParams = {},
): Promise<RunComparisonResponse> {
  return apiRequest<RunComparisonResponse>("/evaluations/compare", {
    query: {
      run_a: runAId,
      run_b: runBId,
      difficulty: params.difficulty ?? undefined,
      tags: params.tags ?? undefined,
      case_status: params.case_status,
      failure_type: params.failure_type ?? undefined,
    },
  });
}

export function buildComparisonExportUrl(
  runAId: string,
  runBId: string,
  format: "csv" | "json" = "csv",
): string {
  const params = new URLSearchParams({
    run_a: runAId,
    run_b: runBId,
    format,
  });
  return `/api/v1/evaluations/compare/export?${params.toString()}`;
}

// ---------------------------------------------------------------------------
// F226 — Benchmark suites and model-profile comparison
// ---------------------------------------------------------------------------

export type BenchmarkSuite = {
  suite_id: string;
  name: string;
  description: string;
  quality_dimension: string;
  case_count: number;
};

export type BenchmarkSuiteListResponse = {
  items: BenchmarkSuite[];
  total: number;
};

export type TriggerBenchmarkRunRequest = {
  suite_id: string;
  provider_profile: "cloud_baseline" | "local_profile" | "fallback_profile";
  evaluation_set_id?: string | null;
  top_k?: number;
  rerank?: boolean;
};

export type TriggerBenchmarkRunResponse = {
  evaluation_run_id: string;
  suite_id: string;
  provider_profile: string;
  status: "queued";
};

export type ProviderProfileSummary = {
  provider_profile: string;
  provider_type: string | null;
  run_count: number;
  latest_run_id: string | null;
  retrieval_hit_rate: number | null;
  citation_accuracy_score: number | null;
  faithfulness_score: number | null;
  answer_relevance_score: number | null;
  not_found_rate: number | null;
  latency_ms_average: number | null;
  cost_usd_total: number | null;
  local_model_metrics: LocalModelMetrics | null;
};

export type ReleaseGateRecommendation = {
  provider_profile: string;
  is_ready: boolean;
  failing_checks: string[];
  passing_checks: string[];
  recommendation: string;
};

export type ModelProfileComparisonReport = {
  organization_id: string;
  evaluation_set_id: string | null;
  profiles: ProviderProfileSummary[];
  release_gate_recommendations: ReleaseGateRecommendation[];
  default_thresholds: Record<string, number>;
  generated_at: string;
};

export async function listBenchmarkSuites(): Promise<BenchmarkSuiteListResponse> {
  return apiRequest<BenchmarkSuiteListResponse>(
    "/evaluations/benchmark-suites",
  );
}

export async function triggerBenchmarkRun(
  suiteId: string,
  payload: TriggerBenchmarkRunRequest,
): Promise<TriggerBenchmarkRunResponse> {
  return apiRequest<TriggerBenchmarkRunResponse>(
    `/evaluations/benchmark-suites/${encodeURIComponent(suiteId)}/run`,
    { method: "POST", json: payload },
  );
}

export async function getModelProfileComparisonReport(
  evaluationSetId?: string | null,
): Promise<ModelProfileComparisonReport> {
  return apiRequest<ModelProfileComparisonReport>(
    "/evaluations/model-profile-report",
    {
      query: evaluationSetId
        ? { evaluation_set_id: evaluationSetId }
        : undefined,
    },
  );
}
